File: examples/LMemLoopback/client/php/BasicStatic/LMemLoopbackClient.php
<?php 
set_include_path(get_include_path() . PATH_SEPARATOR . getenv('PHP_THRIFT_LIB'));

require_once 'Thrift/ClassLoader/ThriftClassLoader.php';

use Thrift\ClassLoader\ThriftClassLoader;

$GEN_DIR = '../gen-php';

$loader = new ThriftClassLoader();
$loader->registerNamespace('Thrift', getenv('PHP_THRIFT_LIB'));
$loader->registerDefinition('com\maxeler\LMemLoopback', $GEN_DIR);
$loader->register();

use Thrift\Protocol\TBinaryProtocol;
use Thrift\Transport\TSocket;
use Thrift\Transport\TBufferedTransport;
use Thrift\Exception\TException;

use com\maxeler\LMemLoopback\LMemLoopbackServiceClient;

function check($dataOutDFE, $dataOutCPU, $size) {
    $status = 0;

    for($i=0; $i < $size; $i++) {
        if($dataOutDFE[$i] != $dataOutCPU[$i]) {
            echo "Output data @ ".$i." = ".$dataOutDFE[$i]." (expected ".$dataOutCPU[$i].")\r\n";
            $status++;
        }
    }
        return $status;
}

function LMemLoopbackCPU ($size, $inA, $inB) {
    for ($i = 0 ; $i < $size ; $i++) {
        $dataOut[$i] = $inA[$i] + $inB[$i];
    }

    return $dataOut;
}

function LMemLoopbackDFE($size, $inA, $inB) {
    $sizeBytes = $size * 4;
    try {
	$startTime = microtime(true);
	# Make socket
        $socket = new TSocket('localhost', 9090);

	# Buffering is critical. Raw sockets are very slow
        $transport = new TBufferedTransport($socket);

	# Wrap in a protocol
        $protocol = new TBinaryProtocol($transport);

	# Create a client to use the protocol encoder
        $client = new LMemLoopbackServiceClient($protocol);

	echo "Creating a client:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Connect!
        $startTime = microtime(true);
        $transport->open();
	echo "Opening connection:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Allocate and send input streams to server
        $startTime = microtime(true);
        $address_inA = $client->malloc_int32_t($size);
        $client->send_data_int32_t($address_inA, $inA);

        $address_inB = $client->malloc_int32_t($size);
        $client->send_data_int32_t($address_inB, $inB);
	echo "Sending input data:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Allocate memory for output stream on server
        $startTime = microtime(true);
        $address_outData = $client->malloc_int32_t($size);
	echo "Allocating memory for output stream on server:\t".(microtime(true) - $startTime)."s\r\n";

        # Writing to LMem
        $startTime = microtime(true);
        $client->LMemLoopback_writeLMem(0, $sizeBytes, $address_inA);
        $client->LMemLoopback_writeLMem($sizeBytes, $sizeBytes, $address_inB);
	echo "Writing to LMem:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Default action
        $startTime = microtime(true);
        $client->LMemLoopback($size);
	echo "LMemLoopback time:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Reading from LMem
        $startTime = microtime(true);
        $client->LMemLoopback_readLMem(2 * $sizeBytes, $sizeBytes, $address_outData);
	echo "Reading from LMem:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Get output stream from server
        $startTime = microtime(true);
        $outData = $client->receive_data_int32_t($address_outData, $size);
	echo "Getting output stream:\t(size = ".($size * 32)." bit)\t".(microtime(true) - $startTime)."s\r\n";

        # Free allocated memory for streams on server
        $startTime = microtime(true);
        $client->free($address_inA);
        $client->free($address_inB);
        $client->free($address_outData);
	echo "Freeing allocated memory for streams on server:\t".(microtime(true) - $startTime)."s\r\n";

        # Close!
        $startTime = microtime(true);
        $transport->close();
	echo "Closing connection:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

    } catch (TException $tx) {
        echo "ThriftException: ".$tx->getMessage()."\r\n";
        exit(-1);
    }
    
    return $outData;
}

# Generate input data
$startTime = microtime(true);
$size = 384;

for($i = 0; $i < $size; $i++) {
    $inA[$i] = $i;
    $inB[$i] = $size - $i;
}
echo "Generating input data:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

# DFE Output
$startTime = microtime(true);
$dataOutDfe = LMemLoopbackDFE($size, $inA, $inB);
echo "LMemLoopback DFE total time:\t\t\t".(microtime(true) - $startTime)."s\r\n";

# CPU Output
$startTime = microtime(true);
$dataOutCpu = LMemLoopbackCPU($size, $inA, $inB);
echo "LMemLoopback CPU total time:\t\t\t".(microtime(true) - $startTime)."s\r\n";

# Checking results
$startTime = microtime(true);
$status = check($dataOutDfe, $dataOutCpu, $size);
echo "Checking results:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

if ($status == 0) { 
    echo "Test successful!\r\n";
} else {
    echo "Test failed ", $status, " times!\r\n";
    exit (-1);
}

File: examples/LMemLoopback/client/perl/BasicStatic/LMemLoopbackClient.pl
#!/usr/bin/perl -w

use strict;
use warnings;

use lib '../gen-perl';

use Thrift;
use Thrift::BinaryProtocol;
use Thrift::Socket;
use Thrift::BufferedTransport;

use com::maxeler::LMemLoopback::LMemLoopbackService;
use com::maxeler::LMemLoopback::Types;

use Data::Dumper;

use Time::HiRes qw/ time /;

sub check {
    my($dataOutDFE, $dataOutCPU, $size) = @_;
    my $status = 0;

    for(my $i = 0; $i < $size; $i++) {
        if($$dataOutDFE[$i] != $$dataOutCPU[$i]) {
            print "Output data @ ".$i." = ".$$dataOutDFE[$i]." (expected ".$$dataOutCPU[$i].")\r\n";
            $status++;
        }
    }
    return $status;    
}

sub LMemLoopbackCPU {
    my ($size, $inA, $inB) = @_;
    my @dataOut = ();

    for (my $i = 0; $i < $size; $i++) {
        $dataOut[$i] = $$inA[$i] + $$inB[$i];
    }

    return \@dataOut;
}

sub LMemLoopbackDFE {
    my ($size, $inA, $inB) = @_;
    my @outData = ();
    my $sizeBytes = $size * 4;
    eval {
	my $startTime = time;

        # Make socket
        my $socket    = new Thrift::Socket('localhost', 9090);

        # Buffering is critical. Raw sockets are very slow
        my $transport = new Thrift::BufferedTransport($socket);

        # Wrap in a protocol
        my $protocol  = new Thrift::BinaryProtocol($transport);

        # Create a client to use the protocol encoder
        my $client    = new com::maxeler::LMemLoopback::LMemLoopbackServiceClient($protocol);
	print "Creating a client:\t\t\t\t", (time - $startTime), "s\n";

        # Connect!
	$startTime = time;
        $transport->open();
	print "Opening connection:\t\t\t\t", (time - $startTime), "s\n";

        # Allocate and send input streams to server
	$startTime = time;
        my $address_inA = $client->malloc_int32_t($size);
        $client->send_data_int32_t($address_inA, $inA);

        my $address_inB = $client->malloc_int32_t($size);
        $client->send_data_int32_t($address_inB, $inB);
	print "Sending input data:\t\t\t\t", (time - $startTime), "s\n";

        # Allocate memory for output stream on server
	$startTime = time;
        my $address_outData = $client->malloc_int32_t($size);
        print "Allocating memory for output stream on server:\t", (time - $startTime), "s\n";

        # Write to LMem
	$startTime = time;
        $client->LMemLoopback_writeLMem(0, $sizeBytes, $address_inA);
        $client->LMemLoopback_writeLMem($sizeBytes, $sizeBytes, $address_inB);
	print "Writing to LMem:\t\t\t\t", (time - $startTime), "s\n";

        # Action default
	$startTime = time;
        $client->LMemLoopback($size);
	print "LMemLoopback time:\t\t\t\t", (time - $startTime), "s\n";

        # Read from LMem
	$startTime = time;
        $client->LMemLoopback_readLMem(2 * $sizeBytes, $sizeBytes, $address_outData);
	print "Reading from LMem:\t\t\t\t", (time - $startTime), "s\n";

        # Get output stream from server
	$startTime = time;
        my $outData_ref = $client->receive_data_int32_t($address_outData, $size);
        @outData = @$outData_ref;
	print "Getting output stream:\t(size = ", ($size * 32), " bit)\t", (time - $startTime), "s\n";

        # Free allocated memory for streams on server
	$startTime = time;
        $client->free($address_inA);
        $client->free($address_inB);
        $client->free($address_outData);
	print "Freeing allocated memory for streams on server:\t", (time - $startTime), "s\n";

        # Close!
	$startTime = time;
        $transport->close();
	print "Closing connection:\t\t\t\t", (time - $startTime), "s\n";

    }; if($@){
        warn(Dumper($@));
    }

    return \@outData;
}
my $startTime = time;
my $size = 384;
my @inA = ();
my @inB = ();
my @dataOutCPU = ();
my @dataOutDFE = ();

for(my $i = 0; $i < $size; $i++) {
    $inA[$i] = $i;
    $inB[$i] = $size - $i;
}
print "Generating input data:\t\t\t\t", (time - $startTime), "s\n";

# DFE Output
$startTime = time;
my $dataOutDFE_ref = LMemLoopbackDFE($size, \@inA, \@inB);
print "DFE LMemLoopback total time:\t\t\t", (time - $startTime), "s\n";

# CPU Output
$startTime = time;
my $dataOutCPU_ref = LMemLoopbackCPU($size, \@inA, \@inB);
print "CPU LMemLoopback total time:\t\t\t", (time - $startTime), "s\n";

@dataOutCPU = @$dataOutCPU_ref;
@dataOutDFE = @$dataOutDFE_ref;

# Checking results
$startTime = time;
my $status = check(\@dataOutDFE, \@dataOutCPU, $size);
print "Checking results:\t\t\t\t", (time - $startTime), "s\n";

if ($status == 0) {
    print "Test successful!\n";
} else {
        print "Test failed ", $status, " times!\n";
        exit -1;

}
 
